File: decode_bin.m
function adcp=read_vadcp(fname)

%start by getting first data batch
CHECK = 0;
tic;
fid = fopen(fname);
bufferSize = 1e6;
eoe = 127;

dataBatch = fread(fid,bufferSize,'uint8');

dataIncrement = fread(fid,2,'uint8');

while ~isempty(dataIncrement) && dataIncrement(end) ~= eoe && dataIncrement(end-1) ~= eoe && ~feof(fid)
    dataIncrement(end+1) = fread(fid,1,'uint8');  %This can be slightly optimized
end

data = [dataBatch; dataIncrement(1:end-1)];

cum_index=0;
trip=0;

while ~isempty(data)
    
test=dec2hex(data);
tt=cellstr(test);
I=find(strcmp(tt,'7F'));
J= diff(I)==1;
add_index=cum_index+1+I(J);
if ~exist('ens_index')
    ens_index=I(J);
else
    ens_index=[ens_index; add_index];
end

cum_index=cum_index+length(data);

disp([num2str(length(data)),' bytes scanned with ',num2str(length(I(J))),' ensembles.']);

clear I J tt test add_index;

dataBatch = fread(fid,bufferSize,'uint8');

dataIncrement = fread(fid,2,'uint8');

while ~isempty(dataIncrement) && dataIncrement(end-1) ~= eoe && dataIncrement(end) ~= eoe && ~feof(fid)
    dataIncrement(end+1) = fread(fid,1,'uint8');  %This can be slightly optimized
end

data = [dataBatch; dataIncrement];

%trip=trip+1

end

%%

disp([num2str(length(ens_index)),' ensembles found...processing...']);

fseek(fid,0,'bof');

for ii=1:length(ens_index)-1
    
    
    ens_offset=ens_index(ii)-1;   
    
    if rem(ii,1000)==0
        disp([num2str(floor(ii./length(ens_index).*100)), '% done...']);
    end
    
    
%header data
    offset=ens_offset;
    fseek(fid,offset,'bof');
    cfgid=fread(fid,2,'uint8');
    %fprintf('%s%d%s%s%s\n','index = ',ii,', ID : ', dec2hex(cfgid(2),2),dec2hex(cfgid(1),2));
    hdr.numbytes(ii)=fread(fid,1,'int16');
            
    cmpt=0;
    
    if ii<=1
        cmpt=1;
        test=hdr.numbytes(ii);
    elseif (hdr.numbytes(ii)>=test-10 && hdr.numbytes(ii)<=test+10)
        cmpt=1;
    end
    
    if cmpt==1
    
    fseek(fid,1,'cof');
    hdr.datatypes=fread(fid,1,'int8');
    hdr.dat_offsets=fread(fid,hdr.datatypes,'int16');
    
%       for ii=1:length(hdr.dat_offsets)
%           hdr.dat_offsets(ii)
%       end
    
%configuration data

    offset=ens_offset+hdr.dat_offsets(1);
    fseek(fid,offset,'bof');
    cfgid=fread(fid,2,'uint8');
    %fprintf('%s%s%s\n','ID : ', dec2hex(cfgid(2),2),dec2hex(cfgid(1),2));
    cfg.name           ='vadcp';
    cfg.sourceprog     ='instrument';
    cfg.prog_ver       =fread(fid,1,'uint8')+fread(fid,1,'uint8')/100; 
    config             =fread(fid,2,'uint8');  % Coded stuff
    cfg.config         =[dec2base(config(2),2,8) '-' dec2base(config(1),2,8)];
%     cfg.beam_angle     =getopt(bitand(config(2),3),15,20,30);
%     cfg.numbeams       =getopt(bitand(config(2),16)==16,4,5);
%     cfg.beam_freq      =getopt(bitand(config(1),7),75,150,300,600,1200,2400,38);
%     cfg.beam_pattern   =getopt(bitand(config(1),8)==8,'concave','convex'); % 1=convex,0=concave
%     cfg.orientation    =getopt(bitand(config(1),128)==128,'down','up');    % 1=up,0=down
    fseek(fid,2,'cof');
    cfg.n_beams        =fread(fid,1,'uint8');
    cfg.n_cells        =fread(fid,1,'uint8');
    cfg.pings_per_ensemble=fread(fid,1,'uint16');
    cfg.cell_size      =fread(fid,1,'uint16')*.01;	 % meters
    cfg.blank          =fread(fid,1,'uint16')*.01;	 % meters
    fseek(fid,1,'cof');          %
    cfg.corr_threshold =fread(fid,1,'uint8');
    cfg.n_codereps     =fread(fid,1,'uint8');
    fseek(fid,1,'cof');
    cfg.evel_threshold =fread(fid,1,'uint16');
    cfg.time_between_ping_groups=sum(fread(fid,3,'uint8').*[60 1 .01]'); % seconds
    coord_sys      =fread(fid,1,'uint8');                                % Lots of bit-mapped info
      cfg.coord=dec2base(coord_sys,2,8);
%       cfg.coord_sys      =getopt(bitand(bitshift(coord_sys,-3),3),'beam','instrument','ship','earth');
%       cfg.use_pitchroll  =getopt(bitand(coord_sys,4)==4,'no','yes');  
%       cfg.use_3beam      =getopt(bitand(coord_sys,2)==2,'no','yes');
%       cfg.bin_mapping    =getopt(bitand(coord_sys,1)==1,'no','yes');
    fseek(fid,4,'cof');
    cfg.sensors_src    =dec2base(fread(fid,1,'uint8'),2,8);
    cfg.sensors_avail  =dec2base(fread(fid,1,'uint8'),2,8);
    cfg.bin1_dist      =fread(fid,1,'uint16')*.01;	% meters
    fseek(fid,4,'cof');
    cfg.fls_target_threshold =fread(fid,1,'uint8');
    fseek(fid,1,'cof');
    cfg.xmit_lag       =fread(fid,1,'uint16')*.01; % meters
    fseek(fid,8,'cof');
    cfg.bandwidth     =fread(fid,1,'uint16');
    cfg.syspower     =fread(fid,1,'uint8');
    fseek(fid,1,'cof');
    cfg.sernum     =fread(fid,1,'uint32');
    cfg.b_angle     =fread(fid,1,'uint8');
    
    cfg.ranges=cfg.bin1_dist+[0:(cfg.n_cells-1)]'*cfg.cell_size;
 %   if cfg.orientation==1, cfg.ranges=-cfg.ranges; end   
    
%system data
    offset=ens_offset+hdr.dat_offsets(2);
    fseek(fid,offset,'bof');
    cfgid=fread(fid,2,'uint8');
    %fprintf('%s%s%s\n','ID : ', dec2hex(cfgid(2),2),dec2hex(cfgid(1),2));
    ens.number(ii)         =fread(fid,1,'uint16');
    rtc=fread(fid,7,'uint8');
    ens.mtime(ii)=datenum(datenum(rtc(1)+2000,rtc(2),rtc(3),rtc(4),rtc(5),rtc(6)));
    fseek(fid,5,'cof');
    ens.depth(ii)          =fread(fid,1,'uint16')*.1;   % meters
    fseek(fid,6,'cof');
    ens.salinity(ii)       =fread(fid,1,'int16');       % PSU
    ens.temperature(ii)    =fread(fid,1,'int16')*.01;   % Deg C
    ens.mpt(ii)            =sum(fread(fid,3,'uint8').*[60 1 .01]'); % seconds
    fseek(fid,9,'cof');
    ens.voltage(ii)        =fread(fid,1,'int8').*157/1000;
    
%velocity data
     offset=ens_offset+hdr.dat_offsets(3);
     fseek(fid,offset,'bof');
     cfgid=fread(fid,2,'uint8');
     %fprintf('%s%s%s\n','ID : ', dec2hex(cfgid(2),2),dec2hex(cfgid(1),2));
     vels=fread(fid,[4 cfg.n_cells],'int16')'*.001;     % m/s
     ens.x_vel(:,ii)  =vels(:,1);
     ens.y_vel(:,ii)  =vels(:,2);
     ens.z_vel(:,ii)  =vels(:,3);
     
%correlations
     offset=ens_offset+hdr.dat_offsets(5);
     fseek(fid,offset,'bof');
     cfgid=fread(fid,2,'uint8');
     %fprintf('%s%s%s\n','ID : ', dec2hex(cfgid(2),2),dec2hex(cfgid(1),2));
     temp_corr   =fread(fid,[4 cfg.n_cells],'uint8')';
     ens.corr(:,:,ii)    =temp_corr(:,1:3);
     fseek(fid,1,'cof');

%echo intensity
     offset=ens_offset+hdr.dat_offsets(6);
     fseek(fid,offset,'bof');
     cfgid=fread(fid,2,'uint8');
     %fprintf('%s%s%s\n','ID : ', dec2hex(cfgid(2),2),dec2hex(cfgid(1),2));
     temp_intens   =fread(fid,[4 cfg.n_cells],'uint8')';
     ens.intens(:,:,ii)    =temp_intens(:,1:3);
     
%percent good
     offset=hdr.dat_offsets(7);
     fseek(fid,offset,'bof');
     cfgid=fread(fid,2,'uint8');
     %fprintf('%s%s%s\n','ID : ', dec2hex(cfgid(2),2),dec2hex(cfgid(1),2));
     temp_pg   =fread(fid,[4 cfg.n_cells],'uint8')';
     ens.perc_good(:,:,ii)    =temp_pg(:,1:3);
     
%surface track status output
     offset=ens_offset+hdr.dat_offsets(11);
     fseek(fid,offset,'bof');
     cfgid=fread(fid,2,'uint8');
     %fprintf('%s%s%s\n','ID : ', dec2hex(cfgid(2),2),dec2hex(cfgid(1),2));
     ens.surface_track(ii)=fread(fid,1,'uint32').*.0001; %in m
     ens.surface_track_uncorr(ii)=fread(fid,1,'uint32').*.0001; %in m
     fseek(fid,1,'cof');
     ens.v_amp(ii)=fread(fid,1,'uint8'); %in counts
     ens.v_pgood(ii)=fread(fid,1,'uint8'); %in counts
    
   else
    
     ens.number(ii)         =NaN;
     ens.mtime(ii)          =NaN;
     ens.depth(ii)          =NaN;
     ens.salinity(ii)       =NaN;
     ens.temperature(ii)    =NaN;
     ens.mpt(ii)            =NaN;
     ens.pressure(ii)       =NaN;
     ens.x_vel(:,ii)        =NaN;
     ens.y_vel(:,ii)        =NaN;
     ens.z_vel(:,ii)        =NaN;
     ens.corr(:,:,ii)       =NaN;
     ens.intens(:,:,ii)     =NaN;    
    end
end


adcp=ens;
adcp.config=cfg;

toc

end


%-------------------------------------
function opt=getopt(val,varargin)
% Returns one of a list (0=first in varargin, etc.)
if val+1>length(varargin),
	opt='unknown';
else
   opt=varargin{val+1};
end;
end

File: decode_bin.py
import argparse
import os
import struct
import mmap

import h5py
from tqdm import tqdm

from ensembles import Config, Ensemble, EnsembleFormatError, EnsembleWriter

# Variable for batch writing size. Value was determined by benchmarking.
# With 4096 max sized ensembles that would be ~3MB of data in RAM, which any modern system should be capable of
batch_size = 4096


# Convert file to readable data
def decode_bin(path):
    file = open(path, "rb")
    # Get file size for progress bar
    file_size = os.path.getsize(path)
    
    # Create a memory map for faster access than just the file
    mm = mmap.mmap(file.fileno(), length=0, access=mmap.ACCESS_READ)
    file.close()

    # List for storing start index of each ensamble
    ens_indexes = []
    current_offset = 0

    # Set up progress bar
    scan_progress = tqdm(
        total=file_size,
        unit="B",
        unit_scale=True,
        desc="Scanning file"
    )

    # Search the entire file for ensemble headers
    while current_offset < file_size:
        # not enough bytes left for a header
        if current_offset + 4 > file_size:
                break
        
        # Get header ID, source ID, and numbytes
        header = mm[current_offset:current_offset + 4]

        # If the two start bytes are right (two 7f bytes in a row)
        if header[0] == 0x7f and header[1] == 0x7f:
            # Get ensemble size
            ens_size = header[2] + (header[3] << 8) + 2

            # If size is within expected bounds (removes potential errors from random 7f7f data)
            if 32 <= ens_size <= 4096 and current_offset + ens_size <= file_size:
                ens_indexes.append(current_offset)
                current_offset += ens_size
                scan_progress.update(ens_size)
                continue

        # Continue seeking
        current_offset += 1
        scan_progress.update(1)

    # Needed to finish scan bar since never actually reaches last byte of file
    scan_progress.n = scan_progress.total + 1
    scan_progress.close()

    # Progress bar for decoding and writing batches
    decode_progress = tqdm(
        total=len(ens_indexes),
        unit=" Ensembles",
        desc="Decoding Ensembles",
    )

    batch_progress = tqdm(
        total=len(ens_indexes) // batch_size + 1,
        unit=" Batches",
        desc="Writing Batches"
    )

    # Get the length of an ensemble
    offset = ens_indexes[0]
    numbytes = struct.unpack("<h", mm[offset + 2:offset + 4])[0]

    # Decode the first ensemble to get config and structure data
    ens_dat = mm[offset:offset + numbytes]
    ens = Ensemble.from_bytes(ens_dat)
    decode_progress.update(1)

    if not Ensemble.config:
        raise EnsembleFormatError(
            "Configuration data missing from first ensemble")
    cfg = Ensemble.config

    # Needed for velocity shapes
    n_cells = Ensemble.config.n_cells

    fname = f"{path[:-4]}.hdf5"

    # Set up structure of HDF5 file
    config_file(fname, n_cells, len(ens_indexes), cfg)

    # List for storing ensembles, add first ensemble
    batch = []
    batch.append(ens)

    writer = EnsembleWriter(fname, batch_size, n_cells)

    with h5py.File(fname, "a") as f:
        writer.initialize_datasets(f)
        # For all ensembles, create object and write batches to file
        skipped = 0
        for i in range(1, len(ens_indexes)):
            try:
                ens_dat = mm[ens_indexes[i]:ens_indexes[i] + numbytes]
                ens = Ensemble.from_bytes(ens_dat)
                batch.append(ens)
            except (EnsembleFormatError, IndexError, ValueError, struct.error) as e:
                skipped += 1
                # You can uncomment the next line for debugging:
                # tqdm.write(f"Skipped ensemble at index {i}: {e}")
                continue
            finally:
                decode_progress.update(1)

            if len(batch) == batch_size:
                writer.write_batch(batch, f)
                batch[:] = []
                batch_progress.update(1)

        # Write any leftovers
        if batch:
            writer.write_batch(batch, f)
            batch[:] = []
            batch_progress.update(1)

        decode_progress.set_description(f"Decoding Ensembles (Skipped {skipped})")

    # Extra updates to force bars to render properly
    decode_progress.update(1)
    batch_progress.update(1)
    decode_progress.close()
    batch_progress.close()
    
    mm.close()


def config_file(fname: str, n_cells: int, n_ens: int, cfg: Config):
    """Helper function for creating file to clean up main function a bit"""
    with h5py.File(fname, "w") as f:
        # Config data
        cfg_group = f.create_group("config")
        cfg_group.attrs.create("name", data="vadcp")
        cfg_group.attrs.create("sourceprog", data="instrument")
        cfg_group.attrs.create("prog_ver", data=cfg.prog_ver)
        cfg_group.attrs.create("config", data=cfg.config)
        cfg_group.attrs.create("n_beams", data=cfg.n_beams)
        cfg_group.attrs.create("n_cells", data=cfg.n_cells)
        cfg_group.attrs.create("pings_per_ensemble",
                               data=cfg.pings_per_ensemble)
        cfg_group.attrs.create("cell_size", data=cfg.cell_size)
        cfg_group.attrs.create("blank", data=cfg.blank)
        cfg_group.attrs.create("corr_threshold", data=cfg.corr_threshold)
        cfg_group.attrs.create("n_codereps", data=cfg.n_codereps)
        cfg_group.attrs.create("evel_threshold", data=cfg.evel_threshold)
        cfg_group.attrs.create("time_between_ping_groups",
                               data=cfg.time_between_ping_groups)
        cfg_group.attrs.create("coord", data=cfg.coord)
        cfg_group.attrs.create("sensors_src", data=cfg.sensors_src)
        cfg_group.attrs.create("sensors_avail", data=cfg.sensors_avail)
        cfg_group.attrs.create("bin1_dist", data=cfg.bin1_dist)
        cfg_group.attrs.create("fls_target_threshold",
                               data=cfg.fls_target_threshold)
        cfg_group.attrs.create("xmit_lag", data=cfg.xmit_lag)
        cfg_group.attrs.create("bandwidth", data=cfg.bandwidth)
        cfg_group.attrs.create("syspower", data=cfg.syspower)
        cfg_group.attrs.create("sernum", data=cfg.sernum)
        cfg_group.create_dataset("ranges", data=cfg.ranges, dtype="float")

        # Ensemble data
        ens_group = f.create_group("ensembles")

        ds = ens_group.create_dataset("number", shape=(
            n_ens,), dtype="uint32", chunks=batch_size, maxshape=(None,), compression="gzip")
        add_metadata(ds, units="1", long_name="Ensemble Number")

        ds = ens_group.create_dataset("mtime", shape=(
            n_ens,), dtype="uint32", chunks=batch_size, maxshape=(None,), compression="gzip")
        add_metadata(ds, units="seconds", long_name="Time")

        ds = ens_group.create_dataset("depth", shape=(
            n_ens,), dtype="uint16", chunks=batch_size, maxshape=(None,), compression="gzip")
        add_metadata(ds, units="meters",
                     long_name="Transducer Depth", scale_factor=0.1)

        ds = ens_group.create_dataset("salinity", shape=(
            n_ens,), dtype="int16", chunks=batch_size, maxshape=(None,), compression="gzip")
        add_metadata(ds, units="ppt", long_name="Water Salinity")

        ds = ens_group.create_dataset("temperature", shape=(
            n_ens,), dtype="int16", chunks=batch_size, maxshape=(None,), compression="gzip")
        add_metadata(ds, units="degrees C",
                     long_name="Water Temperature", scale_factor=0.01)

        ds = ens_group.create_dataset("mpt", shape=(
            n_ens,), dtype="uint32", chunks=batch_size, maxshape=(None,), compression="gzip")
        add_metadata(ds, units="seconds",
                     long_name="Sleep Duration", scale_factor=0.01)

        ds = ens_group.create_dataset("voltage", shape=(
            n_ens,), dtype="uint8", chunks=batch_size, maxshape=(None,), compression="gzip")
        add_metadata(ds, units="volts",
                     long_name="Battery Voltage", scale_factor=0.157)

        ds = ens_group.create_dataset("x_vel", shape=(n_ens, n_cells), dtype="int16", chunks=(
            batch_size, n_cells), maxshape=(None, n_cells), compression="gzip")
        add_metadata(ds, units="m/s",
                     long_name="X Horizontal Velocity", scale_factor=0.001)

        ds = ens_group.create_dataset("y_vel", shape=(n_ens, n_cells), dtype="int16", chunks=(
            batch_size, n_cells), maxshape=(None, n_cells), compression="gzip")
        add_metadata(ds, units="m/s",
                     long_name="Y Horizontal Velocity", scale_factor=0.001)

        ds = ens_group.create_dataset("z_vel", shape=(n_ens, n_cells), dtype="int16", chunks=(
            batch_size, n_cells), maxshape=(None, n_cells), compression="gzip")
        add_metadata(ds, units="m/s",
                     long_name="Z Vertical Velocity", scale_factor=0.001)

        ds = ens_group.create_dataset("corr", shape=(n_ens, n_cells, 3), dtype="uint8", chunks=(
            batch_size, n_cells, 3), maxshape=(None, n_cells, 3), compression="gzip")
        add_metadata(ds, units="1", long_name="Correlation Magnitude")

        ds = ens_group.create_dataset("intens", shape=(n_ens, n_cells, 3), dtype="uint8", chunks=(
            batch_size, n_cells, 3), maxshape=(None, n_cells, 3), compression="gzip")
        add_metadata(ds, units="1", long_name="Echo Intensity")

        ds = ens_group.create_dataset("perc_good", shape=(n_ens, n_cells, 3), dtype="uint8", chunks=(
            batch_size, n_cells, 3), maxshape=(None, n_cells, 3), compression="gzip")
        add_metadata(ds, units="percent", long_name="Percentage of Good Pings")

        ds = ens_group.create_dataset("surface_track", shape=(
            n_ens,), dtype="uint32", chunks=batch_size, maxshape=(None,), compression="gzip")
        add_metadata(ds, units="meters",
                     long_name="Corrected Depth from Surface Track", scale_factor=0.0001)

        ds = ens_group.create_dataset("surface_track_uncorr", shape=(
            n_ens,), dtype="uint32", chunks=batch_size, maxshape=(None,), compression="gzip")
        add_metadata(ds, units="meters",
                     long_name="Uncorrected Depth from Surface Track", scale_factor=0.0001)

        ds = ens_group.create_dataset("v_amp", shape=(
            n_ens,), dtype="uint8", chunks=batch_size, maxshape=(None,), compression="gzip")
        add_metadata(ds, units="1", long_name="Signal Amplitude at Surface")

        ds = ens_group.create_dataset("v_pgood", shape=(
            n_ens,), dtype="uint8", chunks=batch_size, maxshape=(None,), compression="gzip")
        add_metadata(ds, units="percent",
                     long_name="Percentage Good of Surface Track")


def add_metadata(
    dataset: h5py.Dataset,
    units: str | None = None,
    long_name: str | None = None,
    scale_factor: float | None = None
):
    """Helper function for adding metadata to datasets"""
    if units:
        dataset.attrs["units"] = units
    if long_name:
        dataset.attrs["long_name"] = long_name
    if scale_factor:
        dataset.attrs["scale_factor"] = scale_factor


def main():
    # Command line argument for file path
    parser = argparse.ArgumentParser(
        prog="DecodeADCP",
        description="Parses VADCP output data and decodes it"
    )

    parser.add_argument("path", help="Path to data file")
    args = parser.parse_args()

    # Get path from arguments, make sure file exists
    path = args.path
    if not os.path.isfile(path):
        raise FileNotFoundError(f"Invalid Path Provided {path}")

    decode_bin(path)


if __name__ == "__main__":
    main()
